internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
 // ChunkServiceHandler struct implements the ChunkService interface
 type ChunkServiceHandler struct {
 	locks *common.LockMap
-	// more fields can be added here as needed
+
+	// map to store the rvID to rvInfo mapping
+	// rvIDMap contains any and all cluster awareness information that a node needs to have,
+	// things like what all RVs are hosted by this node, state of each of those RVs,
+	// what all MVs are hosted by these RVs, state of those MVs, etc.
+	// It is initially created from the clustermap which is the source of truth regarding cluster information,
+	// and once the cluster is working it's updated using various RPCs.
+	// Note that any time Cluster Manager needs to update clustermap, before publishing the updated clustermap,
+	// it'll send our one or more RPCs to update the rvIDMap info in all the affected nodes,
+	// thus rvIDMap always contains the latest info and hence is used by RVs to fail requests
+	// which might be sent by nodes having a stale clustermap.
+	rvIDMap map[string]*rvInfo
+}
+
+type rvInfo struct {
+	rvID     string       // id for this RV
+	rvName   string       // rv0, rv1, etc.
+	cacheDir string       // cache dir path for this RV
+	mvMap    sync.Map     // sync map of MV id against mvInfo
+	mvCount  atomic.Int64 // count of MVs for this RV, this should be updated whenever a MV is added or removed from the sync map
+}
+
+type mvInfo struct {
+	mu           sync.Mutex
+	mvName       string   // mv0, mv1, etc.
+	componentRVs []string // sorted list of component RVs for this MV
+
+	// count of chunk operations (get, put or remove) for this MV.
+	// This is used to block the end sync call till all the ongoing chunk operations are completed.
+	chunkOpsCnt atomic.Int64
+
+	// flag to block chunk operations (get, put or remove) for this MV.
+	// This flag is enabled in the end sync call to pause new chunk operations in the MV.
+	// When the end sync call is completed, this flag is disabled.
+	blockOpsFlag atomic.Bool
+
```

Comment:
one more useful information to store for an mv is the totalChunkSize i.e., the total amount of space used up inside an mv by all the chunks stored in it. This can be easily updated in PutChunk and DeleteChunk and is useful for finding candidate RVs when we are looking to replace one. We need an RV with at least that much space.
Didn't you need it for JoinMV as it has the "space to be reserved" as one of the params and that comes from this field.

Reply:
In JoinMV, I am calling the cluster manager API/helper method to get the available the disk space for the given cache directory.
Even if we add `totalChunkBytes` field in the `mvInfo`, we will know the total space utilized by the RV (by calculating the sum of all MVs totalChunkBytes value). We would however need the total available space in the RV for it to join a MV. I am in this case relying on the cluster manager to give me that information, as it is already calculating it.

Reply:
this is not for finding RV free space, but it's for finding the space requirement of an MV when we have to find a replacement RV for it, the replacement RV must have this much free space.

Reply:
Added TODO, will add `totalChunkBytes` and `reserveSpace` in another PR